hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveEntityUpdateAction.java

```diff
@@ -135,28 +141,96 @@ private CompletionStage<EntityEntry> handleGeneratedProperties(EntityEntry entry
 				// that uses all the state values in the where clause, the entry state needs to be
 				// updated otherwise the statement execution will not delete any row (see HHH-15218).
-				entry.postUpdate( getInstance(), getState(), getNextVersion() );
+				entry.postUpdate(instance, getState(), getNextVersion() );
 			}
 		}
 	}
 
+	@Override
+	protected void handleNaturalIdResolutions(EntityPersister persister, SharedSessionContractImplementor session, Object id) {
+		NaturalIdMapping naturalIdMapping = getNaturalIdMapping();
+		if ( naturalIdMapping != null ) {
+			session.getPersistenceContextInternal().getNaturalIdResolutions().manageSharedResolution(
+					id,
+					naturalIdMapping.extractNaturalIdFromEntityState( getState() ),
+					getPreviousNaturalIdValues(),
+					persister,
+					CachedNaturalIdValueSource.UPDATE
+			);
+		}
+	}
+
+	@Override
+	protected void updateCacheItem(Object previousVersion, Object ck, EntityEntry entry) {
+		final EntityPersister persister = getPersister();
+		if ( persister.canWriteToCache() ) {
+			final SharedSessionContractImplementor session = getSession();
+			if ( isCacheInvalidationRequired( persister, session ) || entry.getStatus() != Status.MANAGED ) {
+				persister.getCacheAccessStrategy().remove( session, ck );
+			}
+			else if ( session.getCacheMode().isPutEnabled() ) {
+				//TODO: inefficient if that cache is just going to ignore the updated state!
+				final CacheEntry ce = persister.buildCacheEntry( getInstance(), getState(), getNextVersion(), getSession() );
+				setCacheEntry( persister.getCacheEntryStructure().structure( ce ) );
+				final boolean put = updateCache( persister, previousVersion, ck );
+
+				final StatisticsImplementor statistics = session.getFactory().getStatistics();
+				if ( put && statistics.isStatisticsEnabled() ) {
+					statistics.entityCachePut(
+							StatsHelper.INSTANCE.getRootEntityRole(persister),
+							getPersister().getCacheAccessStrategy().getRegion().getName()
+					);
+				}
+			}
+		}
+	}
+
+	private static boolean isCacheInvalidationRequired(
+			EntityPersister persister,
+			SharedSessionContractImplementor session) {
+		// the cache has to be invalidated when CacheMode is equal to GET or IGNORE
+		return persister.isCacheInvalidationRequired()
+			|| session.getCacheMode() == CacheMode.GET
+			|| session.getCacheMode() == CacheMode.IGNORE;
+	}
+
+	@Override
+	protected Object lockCacheItem(Object previousVersion) {
+		final EntityPersister persister = getPersister();
+		if ( persister.canWriteToCache() ) {
+			final SharedSessionContractImplementor session = getSession();
+			final EntityDataAccess cache = persister.getCacheAccessStrategy();
+			final Object ck = cache.generateCacheKey(
+					getId(),
+					persister,
+					session.getFactory(),
+					session.getTenantIdentifier()
+			);
+			setLock( cache.lockItem( session, ck, previousVersion ) );
+			return ck;
+		}
+		else {
+			return null;
+		}
+	}
+
```

Comment:
Why do we suddenly need all this additional code?

Reply:
fixed